mirgecom/steppers.py

```diff
@@ -93,3 +94,186 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
+    """Set up leap advancer for problems not using advance_state.
+
+    Parameters
+    ----------
+    timestepper
+        Leap method that advances the state from t=time to t=(time+dt), and
+        returns the advanced state.
+    component_id
+        State id (required input for leap method generation)
+    rhs
+        Function that should return the time derivative of the state
+    t: float
+        Time at which to start
+    dt: float
+        Initial timestep to be set by leap method
+    state: numpy.ndarray
+        Agglomerated object array containing at least the state variables that
+        will be advanced by this stepper
+
+    Returns
+    -------
+    dagrt.codegen.python.StepperInterface
+        Python class implementing leap method, and generated by dagrt
+    """
+    code = timestepper.generate()
+    from dagrt.codegen import PythonCodeGenerator
+    codegen = PythonCodeGenerator(class_name="Method")
+    stepper_cls = codegen.get_class(code)(function_map={
+        "<func>" + component_id: rhs,
+        })
+    stepper_cls.set_up(t_start=t, dt_start=dt, context={component_id: state})
+
+    return stepper_cls
+
+
+def advance_state(rhs, timestepper, checkpoint, get_timestep, state, t_final,
+                    component_id="state", t=0.0, istep=0, logmgr=None,
+                    eos=None, dim=None):
+    """Wrap function for advance_state and advance_state_leap.
+
+    Parameters
+    ----------
+    rhs
+        Function that should return the time derivative of the state
+    timestepper
+        Function that advances the state from t=time to t=(time+dt), and
+        returns the advanced state. This function checks if `timestepper`
+        is a leap method or not, and calls the corresponding advancer.
+    checkpoint
+        Function is user-defined and can be used to preform simulation status
+        reporting, viz, and restart i/o.  A non-zero return code from this function
+        indicates that this function should stop gracefully.
+    component_id
+        State id (required input for leap method generation)
+    get_timestep
+        Function that should return dt for the next step. This interface allows
+        user-defined adaptive timestepping. A negative return value indicated that
+        the stepper should stop gracefully.
+    state: numpy.ndarray
+        Agglomerated object array containing at least the state variables that
+        will be advanced by this stepper
+    t_final: float
+        Simulated time at which to stop
+    t: float
+        Time at which to start
+    istep: int
+        Step number from which to start
+
+    Returns
+    -------
+    istep: int
+        the current step number
+    t: float
+        the current time
+    state: numpy.ndarray
+    """
+    from mirgecom.integrators import (rk4_step, euler_step,
+                                      lsrk54_step, lsrk144_step)
+    if timestepper in (rk4_step, euler_step, lsrk54_step, lsrk144_step):
```

Comment:
This is too restrictive. A user needs to be able to pass their own function, and this needs to continue to work. You should do this by recognizing leap method generators.